File: backend/app/schemas/team.py
"""
Team schemas for API request/response validation.
"""

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


class CoordinationRules(BaseModel):
    """Team coordination rules configuration."""

    first_speaker: str = Field(
        default="highest_priority",
        description="First speaker selection: 'highest_priority' or specific agent_id",
    )
    turn_taking: str = Field(
        default="round_robin",
        description="Turn taking mode: round_robin, priority_based, free",
    )
    max_rounds: int = Field(
        default=0,
        ge=0,
        description="Maximum discussion rounds (0 = unlimited)",
    )
    termination: dict = Field(
        default_factory=lambda: {"type": "consensus", "consensus_threshold": 0.8},
        description="Termination conditions",
    )


class OutputRules(BaseModel):
    """Team output rules configuration."""

    mode: str = Field(
        default="merged",
        pattern="^(individual|merged|summary)$",
        description="Output mode",
    )
    summary_agent_id: Optional[str] = Field(
        default=None,
        description="Agent ID for summary (if mode is 'summary')",
    )
    format: str = Field(
        default="markdown",
        pattern="^(text|markdown|json)$",
        description="Output format",
    )


class TeamMemberCreate(BaseModel):
    """Schema for adding a member to a team."""

    agent_id: str = Field(..., description="Agent ID to add")
    role_override: Optional[str] = Field(default=None, max_length=100, description="Override role")
    priority_override: Optional[int] = Field(default=None, ge=1, le=10, description="Override priority")
    config_override: dict = Field(default_factory=dict, description="Additional config overrides")
    position: Optional[int] = Field(
        default=None,
        ge=0,
        description="Position in team (for ordered modes); if omitted, appended to the end",
    )


class TeamMemberResponse(BaseModel):
    """Schema for team member response."""

    id: str
    agent_id: str
    role_override: Optional[str]
    priority_override: Optional[int]
    config_override: dict
    position: int
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class TeamBase(BaseModel):
    """Base team schema with common fields."""

    name: str = Field(..., min_length=1, max_length=100, description="Team name")
    description: Optional[str] = Field(default=None, description="Team description")
    icon: Optional[str] = Field(default=None, max_length=500, description="Team icon URL")

    # Collaboration configuration
    collaboration_mode: str = Field(
        default="roundtable",
        pattern="^(roundtable|pipeline|debate|freeform|custom)$",
        description="Collaboration mode",
    )
    mode_config: dict = Field(
        default_factory=dict,
        description="Mode-specific configuration",
    )

    # Coordinator
    coordinator_id: Optional[str] = Field(
        default=None,
        description="Coordinator agent ID (null for system-managed)",
    )

    # Rules
    coordination_rules: CoordinationRules = Field(
        default_factory=CoordinationRules,
        description="Coordination rules",
    )
    output_rules: OutputRules = Field(
        default_factory=OutputRules,
        description="Output rules",
    )


class TeamCreate(TeamBase):
    """Schema for creating a new team."""

    members: list[TeamMemberCreate] = Field(
        default_factory=list,
        description="Initial team members",
    )
    is_template: bool = Field(default=False, description="Whether this is a template")
    is_public: bool = Field(default=False, description="Whether publicly visible")


class TeamUpdate(BaseModel):
    """Schema for updating a team. All fields are optional."""

    name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    description: Optional[str] = None
    icon: Optional[str] = Field(default=None, max_length=500)

    collaboration_mode: Optional[str] = Field(
        default=None,
        pattern="^(roundtable|pipeline|debate|freeform|custom)$",
    )
    mode_config: Optional[dict] = None
    coordinator_id: Optional[str] = None

    coordination_rules: Optional[CoordinationRules] = None
    output_rules: Optional[OutputRules] = None

    is_public: Optional[bool] = None


class TeamResponse(TeamBase):
    """Schema for team response."""

    id: str
    user_id: str
    is_template: bool
    is_public: bool

    # Statistics
    usage_count: int
    rating: float
    rating_count: int

    # Members
    members: list[TeamMemberResponse]

    # Timestamps
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TeamListResponse(BaseModel):
    """Schema for team list response (simplified)."""

    id: str
    name: str
    description: Optional[str]
    icon: Optional[str]
    collaboration_mode: str
    member_count: int = Field(default=0, description="Number of members")
    is_template: bool
    is_public: bool
    usage_count: int
    rating: float
    created_at: datetime

    class Config:
        from_attributes = True

File: backend/src/models/team.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CoordinationRules {
    #[serde(default = "default_first_speaker")]
    pub first_speaker: String,
    #[serde(default = "default_turn_taking")]
    pub turn_taking: String,
    #[serde(default)]
    pub max_rounds: i32,
    #[serde(default = "default_termination")]
    pub termination: Value,
}

impl Default for CoordinationRules {
    fn default() -> Self {
        Self {
            first_speaker: default_first_speaker(),
            turn_taking: default_turn_taking(),
            max_rounds: 0,
            termination: default_termination(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OutputRules {
    #[serde(default = "default_output_mode")]
    pub mode: String,
    #[serde(default)]
    pub summary_agent_id: Option<String>,
    #[serde(default = "default_output_format")]
    pub format: String,
}

impl Default for OutputRules {
    fn default() -> Self {
        Self {
            mode: default_output_mode(),
            summary_agent_id: None,
            format: default_output_format(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamMember {
    pub id: String,
    pub agent_id: String,
    pub role_override: Option<String>,
    pub priority_override: Option<i32>,
    #[serde(default)]
    pub config_override: Value,
    pub position: i32,
    pub is_active: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Team {
    pub id: String,
    pub user_id: String,
    pub name: String,
    pub description: Option<String>,
    pub icon: Option<String>,
    pub collaboration_mode: String,
    #[serde(default)]
    pub mode_config: Value,
    pub coordinator_id: Option<String>,
    #[serde(default)]
    pub coordination_rules: CoordinationRules,
    #[serde(default)]
    pub output_rules: OutputRules,
    pub is_template: bool,
    pub is_public: bool,
    pub usage_count: u32,
    pub rating: f64,
    pub rating_count: u32,
    #[serde(default)]
    pub members: Vec<TeamMember>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamListItem {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub icon: Option<String>,
    pub collaboration_mode: String,
    pub member_count: usize,
    pub is_template: bool,
    pub is_public: bool,
    pub usage_count: u32,
    pub rating: f64,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamMemberCreate {
    pub agent_id: String,
    #[serde(default)]
    pub role_override: Option<String>,
    #[serde(default)]
    pub priority_override: Option<i32>,
    #[serde(default)]
    pub config_override: Value,
    #[serde(default)]
    pub position: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TeamCreate {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub icon: Option<String>,
    #[serde(default = "default_collaboration_mode")]
    pub collaboration_mode: String,
    #[serde(default)]
    pub mode_config: Value,
    #[serde(default)]
    pub coordinator_id: Option<String>,
    #[serde(default)]
    pub coordination_rules: CoordinationRules,
    #[serde(default)]
    pub output_rules: OutputRules,
    #[serde(default)]
    pub members: Vec<TeamMemberCreate>,
    #[serde(default)]
    pub is_template: bool,
    #[serde(default)]
    pub is_public: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct TeamUpdate {
    pub name: Option<String>,
    pub description: Option<String>,
    pub icon: Option<String>,
    pub collaboration_mode: Option<String>,
    pub mode_config: Option<Value>,
    pub coordinator_id: Option<String>,
    pub coordination_rules: Option<CoordinationRules>,
    pub output_rules: Option<OutputRules>,
    pub members: Option<Vec<TeamMemberCreate>>,
    pub is_public: Option<bool>,
}

fn default_first_speaker() -> String {
    "highest_priority".to_string()
}

fn default_turn_taking() -> String {
    "round_robin".to_string()
}

fn default_termination() -> Value {
    serde_json::json!({"type": "consensus", "consensus_threshold": 0.8})
}

fn default_output_mode() -> String {
    "merged".to_string()
}

fn default_output_format() -> String {
    "markdown".to_string()
}

fn default_collaboration_mode() -> String {
    "roundtable".to_string()
}

